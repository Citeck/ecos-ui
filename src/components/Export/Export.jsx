import React, { Component } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import omit from 'lodash/omit';
import get from 'lodash/get';
import isEmpty from 'lodash/isEmpty';
import cloneDeep from 'lodash/cloneDeep';
import queryString from 'query-string';

import { UserConfigApi } from '../../api/userConfig';
import { URL } from '../../constants';
import { ALFRESCO, PROXY_URI } from '../../constants/alfresco';
import { t } from '../../helpers/util';
import { decodeLink } from '../../helpers/urls';
import { Dropdown } from '../common/form';
import { TwoIcoBtn } from '../common/btns';
import { PREDICATE_AND } from '../common/form/SelectJournal/predicates';
import ParserPredicate from '../Filters/predicates/ParserPredicate';

import './Export.scss';

const api = new UserConfigApi();

export default class Export extends Component {
  static propTypes = {
    className: PropTypes.string,
    dashletConfig: PropTypes.object,
    journalConfig: PropTypes.object,
    grid: PropTypes.object,
    right: PropTypes.bool
  };

  static defaultProps = {
    className: ''
  };

  constructor(props) {
    super(props);
    this.textInput = React.createRef();
    this.form = React.createRef();
  }

  state = {
    isOpen: false
  };

  get dropdownSource() {
    return [
      { id: 0, title: t('export-component.action.html-read'), type: 'html', download: false, target: '_blank' },
      { id: 1, title: t('export-component.action.html-load'), type: 'html', download: true, target: '_self' },
      { id: 2, title: 'Excel', type: 'xlsx', download: true, target: '_self' },
      { id: 3, title: 'CSV', type: 'csv', download: true, target: '_self' },
      { id: 4, title: t('export-component.action.copy-link'), click: this.onCopyUrl }
    ];
  }

  get isShow() {
    const sourceId = get(this.props, 'journalConfig.sourceId') || '';
    const [first, second] = sourceId.split('/');

    return !second || first === ALFRESCO;
  }

  getStateOpen = isOpen => {
    this.setState({ isOpen });
  };

  export = item => {
    if (item.target) {
      const { journalConfig, grid } = this.props;
      const query = this.getQuery(journalConfig, item.type, grid);

      this.textInput.current.value = JSON.stringify(query);

      const form = this.form.current;

      form.action = `${PROXY_URI}report/predicate-report?download=${item.download}`;
      form.target = item.target;

      form.submit();
    } else if (typeof item.click === 'function') {
      item.click();
    }
  };

  getSearchPredicate = (grid = {}) => {
    const { search: text, columns, groupBy } = grid;

    if (isEmpty(text)) {
      return {};
    }

    return ParserPredicate.getSearchPredicates({ text, columns, groupBy });
  };

  getQuery = (config, type, grid) => {
    grid = grid || {};
    config = config || {};
    config.meta = config.meta || {};
    config.meta.createVariants = config.meta.createVariants || [];

    const name = (config.meta.createVariants[0] || {}).title || config.meta.title;
    const reportColumns = (grid.columns || config.columns || [])
      .filter(c => c.default)
      .map(column => ({ attribute: column.attribute, title: column.text }));
    const gridPredicate = get(grid, ['predicates', 0], {});
    const mainPredicate = get(grid, 'predicate', {});
    const searchPredicate = this.getSearchPredicate(grid);
    const predicates = [mainPredicate, searchPredicate, gridPredicate];
    const predicate = cloneDeep({ t: PREDICATE_AND, val: predicates });

    const query = {
      sortBy: grid.sortBy || [{ attribute: 'cm:created', order: 'desc' }],
      predicate: ParserPredicate.removeEmptyPredicates([predicate])[0] || null,
      reportType: type,
      reportTitle: name,
      reportColumns: reportColumns,
      reportFilename: `${name}.${type}`
    };

    (config.meta.criteria || []).forEach((criterion, idx) => {
      query['field_' + idx] = criterion.field;
      query['predicate_' + idx] = criterion.predicate;
      query['value_' + idx] = criterion.value;
    });

    return query;
  };

  getSelectionFilter = () => {
    const { columns } = this.props.journalConfig || {};
    const { groupBy, sortBy, pagination, predicates } = this.props.grid || {};

    return { columns, groupBy, sortBy, pagination, predicate: predicates[0] };
  };

  getSelectionUrl = () => {
    const { dashletConfig } = this.props;
    const { href, host } = window.location;

    if (dashletConfig) {
      const { journalId, journalsListId } = dashletConfig;

      return decodeLink(`${host}${URL.JOURNAL}?${queryString.stringify({ journalId, journalsListId })}`);
    }

    const objectUrl = queryString.parseUrl(href);
    const { journalId, journalsListId } = objectUrl.query;

    return `${objectUrl.url}?${queryString.stringify({ journalId, journalsListId })}`;
  };

  onCopyUrl = () => {
    const data = this.getSelectionFilter();
    const url = this.getSelectionUrl();

    api.copyUrlConfig({ data, url });
  };

  render() {
    const { right, className, children, ...props } = this.props;
    const { isOpen } = this.state;
    const attributes = omit(props, ['journalConfig', 'dashletConfig', 'grid']);

    return this.isShow ? (
      <div {...attributes} className={classNames('ecos-btn-export', className)}>
        <Dropdown
          source={this.dropdownSource}
          value={0}
          valueField={'id'}
          titleField={'title'}
          isButton={true}
          onChange={this.export}
          right={right}
          getStateOpen={this.getStateOpen}
        >
          {children || (
            <TwoIcoBtn
              icons={['icon-upload', isOpen ? 'icon-small-up' : 'icon-small-down']}
              className="ecos-btn_grey ecos-btn_settings-down ecos-btn_x-step_10"
            />
          )}
        </Dropdown>

        <form ref={this.form} method="post" encType="multipart/form-data">
          <input ref={this.textInput} type="hidden" name="jsondata" value="" />
        </form>
      </div>
    ) : null;
  }
}
