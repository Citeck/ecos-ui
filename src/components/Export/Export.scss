@import '../../constants';

.export {
  display: inline-block;
}
