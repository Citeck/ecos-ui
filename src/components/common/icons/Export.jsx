import React from 'react';

export default ({ width = 19, height = 19, viewBox = '0 0 19 19', fill = 'none' }) => (
  <svg width={width} height={height} viewBox={viewBox} fill={fill} xmlns="http://www.w3.org/2000/svg">
    <path
      d="M13.4335 6.90812L9.12502 2.59961M9.12502 2.59961L4.81651 6.90812M9.12502 2.59961L9.12502 11.7911"
      stroke="#B7B7B7"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
    />
    <path
      d="M3.09302 13.3906L3.09302 14.2523C3.09302 15.2041 3.86461 15.9757 4.81642 15.9757L13.4334 15.9757C14.3853 15.9757 15.1568 15.2041 15.1568 14.2523V13.3906"
      stroke="#B7B7B7"
      strokeWidth="2"
      strokeLinecap="round"
    />
  </svg>
);
