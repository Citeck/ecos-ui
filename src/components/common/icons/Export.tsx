import React from 'react';

export default ({ width = 19, height = 19, viewBox = '0 0 19 19', fill = 'none' }) => (
  <svg width={width} height={height} viewBox={viewBox} fill={fill} xmlns="http://www.w3.org/2000/svg">
    <path
      d="M5.07568 8.28329L9.38419 12.5918M9.38419 12.5918L13.6927 8.28329M9.38419 12.5918L9.38419 3.40031"
      stroke="#B7B7B7"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
    />
    <path
      d="M3.35229 14.1914L3.35229 15.0531C3.35229 16.0049 4.12389 16.7765 5.0757 16.7765L13.6927 16.7765C14.6445 16.7765 15.4161 16.0049 15.4161 15.0531V14.1914"
      stroke="#B7B7B7"
      strokeWidth="2"
      strokeLinecap="round"
    />
  </svg>
);
