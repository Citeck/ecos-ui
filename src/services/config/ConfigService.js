import _ from 'lodash';

import { URL } from '../../constants';
import { CITECK_URI } from '../../constants/alfresco';
import { getCurrentLocale } from '../../helpers/export/util';
import { IS_TEST_ENV } from '../../helpers/util';

import { loadConfigs, saveConfig } from './configApi';

const LOCAL_STORAGE_KEY = 'ecos-ui-config';

const TYPE_JSON = 'JSON';
const TYPE_TEXT = 'TEXT';
const TYPE_NUMBER = 'NUMBER';
const TYPE_ASSOC = 'ASSOC';
const TYPE_BOOLEAN = 'BOOLEAN';

/**
 * Max cache time determine time diff since last configs update when we should
 * perform force loading. Force loading mean that UI will wait until config will be updated.
 *
 * When user works with system cache will be updated every MIN_CACHE_TIME_MS
 * and force updating should not occur. This will slightly increase page responsiveness.
 */
const MAX_CACHE_TIME_MS = 1800000; // 30 min
const MIN_CACHE_TIME_MS = 10000; // 10 sec

export const ACTIVE_THEME = 'active-theme';
export const TABS_ENABLED = 'tabs-enabled';
export const NEW_JOURNAL_ENABLED = 'new-journal-enabled';
export const DEFAULT_WORKSPACE = 'default-workspace';
export const WORKSPACES_ENABLED = 'workspaces-enabled';
export const ALFRESCO_ENABLED = 'app/alfresco$alfresco-enabled';
export const ALFRESCO_EXPORT_SRC_ID_PATTERN = 'app/alfresco$alfresco-export-src-id-pattern';

export const FOOTER_CONTENT = 'footer-content';

export const MAIN_MENU_TYPE = 'main-menu-type';
export const MENU_GROUP_PRIORITY = 'menu-group-priority';
export const CREATE_MENU_TYPE = 'create-menu-type';

export const ORGSTRUCT_ALL_USERS_GROUP_SHORT_NAME = 'orgstruct-allUsers-group-shortName';
export const ORGSTRUCT_SEARCH_USER_EXTRA_FIELDS = 'orgstruct-search-user-extra-fields';
export const ORGSTRUCT_HIDE_DISABLED_USERS = 'hide-disabled-users-for-everyone';
export const ORGSTRUCT_SEARCH_USER_MIDLLE_NAME = 'orgstruct-search-user-middle-name';
export const ORGSTRUCT_SHOW_INACTIVE_USER_ONLY_FOR_ADMIN = 'orgstruct-show-inactive-user-only-for-admin';
export const HIDE_IN_ORGSTRUCT = 'hide-in-orgstruct';
export const ORGSTRUCT_SHOW_USERNAME_MASK = 'orgstruct-username-mask';

export const RESTRICT_ACCESS_TO_EDIT_DASHBOARD = 'restrict-access-to-edit-dashboard';
export const RESTRICT_ACCESS_TO_EDIT_DASHBOARD_WIDGETS = 'restrict-access-to-edit-dashboard-widgets';
export const HOME_LINK_URL = 'home-link-url';
export const LOGIN_PAGE_REDIRECT_URL = 'login-page-redirect-url';
export const CUSTOM_FEEDBACK_URL = 'custom-feedback-url';
export const CUSTOM_REPORT_ISSUE_URL = 'custom-report-issue-url';
export const SEPARATE_ACTION_LIST_FOR_QUERY = 'separate-action-list-for-query';
export const SITE_DASHBOARD_ENABLE = 'site-dashboard-enable';
export const TOUCH_CONFIG = 'app/gateway$touch';

const CONFIG_PROPS = {
  [MAIN_MENU_TYPE]: {
    defaultValue: 'left-v1'
  },
  [CREATE_MENU_TYPE]: {
    defaultValue: 'cascad'
  },
  [RESTRICT_ACCESS_TO_EDIT_DASHBOARD]: {
    defaultValue: true
  },
  [RESTRICT_ACCESS_TO_EDIT_DASHBOARD_WIDGETS]: {
    default: null
  },
  [ORGSTRUCT_ALL_USERS_GROUP_SHORT_NAME]: {
    defaultValue: 'all'
  },
  [ORGSTRUCT_HIDE_DISABLED_USERS]: {
    type: TYPE_BOOLEAN,
    defaultValue: true
  },
  [ORGSTRUCT_SEARCH_USER_MIDLLE_NAME]: {
    type: TYPE_BOOLEAN,
    defaultValue: false
  },
  [ORGSTRUCT_SHOW_INACTIVE_USER_ONLY_FOR_ADMIN]: {
    type: TYPE_BOOLEAN,
    defaultValue: false
  },
  [HIDE_IN_ORGSTRUCT]: {
    defaultValue: ''
  },
  [ORGSTRUCT_SHOW_USERNAME_MASK]: {
    type: TYPE_TEXT,
    defaultValue: ''
  },
  [ORGSTRUCT_SEARCH_USER_EXTRA_FIELDS]: {
    multiple: true,
    type: TYPE_TEXT
  },
  [ACTIVE_THEME]: {
    defaultValue: 'ecos'
  },
  [FOOTER_CONTENT]: {
    defaultValue: ''
  },
  [HOME_LINK_URL]: {
    defaultValue: URL.DASHBOARD
  },
  [CUSTOM_FEEDBACK_URL]: {
    defaultValue: 'https://www.citeck.ru/feedback'
  },
  [CUSTOM_REPORT_ISSUE_URL]: {
    defaultValue:
      'mailto:support@citeck.ru?subject=Ошибка в работе Citeck ECOS: ' +
      'краткое описание&body=Summary: Короткое описание проблемы (продублировать в теме письма)' +
      '%0A%0ADescription:%0AПожалуйста, детально опишите возникшую проблему, последовательность ' +
      'действий, которая привела к ней. При необходимости приложите скриншоты.'
  },
  [SEPARATE_ACTION_LIST_FOR_QUERY]: {
    defaultValue: false
  },
  [LOGIN_PAGE_REDIRECT_URL]: {
    defaultValue: null,
    type: TYPE_TEXT
  },
  [ALFRESCO_ENABLED]: {
    defaultValue: false,
    type: TYPE_BOOLEAN
  },
  [SITE_DASHBOARD_ENABLE]: {
    defaultValue: false
  },
  [MENU_GROUP_PRIORITY]: {
    multiple: true,
    type: TYPE_JSON
  },
  [TABS_ENABLED]: {
    defaultValue: true
  },
  [NEW_JOURNAL_ENABLED]: {
    defaultValue: false,
    type: TYPE_BOOLEAN
  },
  [TOUCH_CONFIG]: {
    defaultValue: {
      enabled: true,
      uri: `${CITECK_URI}ecos/touch`
    }
  }
};

/**
 * Configuration service with lazy loading.
 *
 * Main purpose of this service is optimization:
 * We will not wait every page update until configs will be loaded
 * 1. When cache should be updated we check how much time has passed since the last update
 *   and if it is not too much then we do update in background without blocking UI.
 * 2. All config requests will be batched in one request
 *
 * Second purpose is encapsulation of config API details.
 * Every part of system can call await ConfigService.getValue("abc")
 * without worrying about how exactly this config will be obtained.
 */
class ConfigService {
  constructor() {
    this._loadConfigsFunc = loadConfigs;
    if (IS_TEST_ENV) {
      this._loadConfigsFunc = async configsMap => {
        const result = {};
        for (let key in configsMap) {
          result[key] = null;
        }
        return result;
      };
    }

    const currentTime = new Date().getTime();

    let lsConfig = JSON.parse(localStorage.getItem(LOCAL_STORAGE_KEY) || '{}');
    this._values = { ...(lsConfig.values || {}) };

    let timeSinceLastUpdate = currentTime - (lsConfig.lastUpdateTime || 0);
    const currentLocale = getCurrentLocale();
    if (lsConfig.locale !== currentLocale) {
      timeSinceLastUpdate = MAX_CACHE_TIME_MS + 1;
    } else {
      for (let configKey in CONFIG_PROPS) {
        if (!this._values.hasOwnProperty(configKey)) {
          // if localStorage not contains some configs then we should update all configs
          timeSinceLastUpdate = MAX_CACHE_TIME_MS + 1;
          this._values = {};
          break;
        }
      }
    }
    if (timeSinceLastUpdate < MIN_CACHE_TIME_MS) {
      return;
    }

    const configsToLoad = this._getConfigsToLoad();

    const loadedValuesPromise = this._loadConfigsFunc(configsToLoad).then(values => {
      const cfgValues = {};
      for (let configKey in values) {
        let value = values[configKey];
        if (this._isEmptyValue(value)) {
          const confProps = this._getConfigProps({
            ...(CONFIG_PROPS[configKey] || {})
          });
          value = confProps.defaultValue;
        }
        cfgValues[configKey] = {
          [configsToLoad[configKey]]: value
        };
      }
      localStorage.setItem(
        LOCAL_STORAGE_KEY,
        JSON.stringify({
          locale: currentLocale,
          values: cfgValues,
          lastUpdateTime: currentTime
        })
      );
      return values;
    });

    if (timeSinceLastUpdate > MAX_CACHE_TIME_MS) {
      for (let configKey in CONFIG_PROPS) {
        this._values[configKey] = loadedValuesPromise.then(values => {
          return this.setValue(configKey, values[configKey], CONFIG_PROPS[configKey]);
        });
      }
    } else {
      loadedValuesPromise.then(values => {
        if (values) {
          for (let configKey in values) {
            this.setValue(configKey, values[configKey], CONFIG_PROPS[configKey]);
          }
        }
      });
    }
  }

  async setServerValue(key, value, props) {
    const prevValue = this._values[key];
    let valueRes = this.setValue(key, value, props);
    try {
      await saveConfig(key, this.setValue(key, value, props));
    } catch (e) {
      this._values[key] = prevValue;
      throw e;
    }
    return valueRes;
  }

  setValue(key, value, props) {
    let newValue;
    const confProps = this._getConfigProps(props);
    if (this._isEmptyValue(value)) {
      newValue = confProps.defaultValue;
    } else {
      newValue = value;
    }
    this._values[key] = {
      ...(this._values[key] || {}),
      [this._getConfigAttribute(confProps)]: newValue
    };
    return newValue;
  }

  _isEmptyValue(value) {
    return value == null || value === '';
  }

  _getConfigsToLoad() {
    const configsToLoad = {};

    for (let configKey in CONFIG_PROPS) {
      const props = this._getConfigProps(CONFIG_PROPS[configKey]);
      configsToLoad[configKey] = this._getConfigAttribute(props);
    }
    return configsToLoad;
  }

  _getConfigProps(props, value) {
    const result = {
      ...(props || {})
    };
    if (result.type === undefined) {
      let valueToEvalType = value;
      if (valueToEvalType == null && result.defaultValue !== undefined) {
        valueToEvalType = result.defaultValue;
      }
      if (valueToEvalType == null || _.isString(valueToEvalType)) {
        result.type = TYPE_TEXT;
      } else if (_.isNumber(valueToEvalType)) {
        result.type = TYPE_NUMBER;
      } else if (_.isObject(valueToEvalType)) {
        result.type = TYPE_JSON;
      } else if (_.isBoolean(valueToEvalType)) {
        result.type = TYPE_BOOLEAN;
      } else {
        result.type = TYPE_TEXT;
      }
    }
    if (result.multiple === undefined) {
      result.multiple = Array.isArray(value);
    }
    if (result.defaultValue === undefined) {
      if (result.multiple) {
        result.defaultValue = [];
      } else if (result.type === TYPE_TEXT) {
        result.defaultValue = '';
      } else if (result.type === TYPE_NUMBER) {
        result.defaultValue = 0;
      } else if (result.type === TYPE_JSON) {
        result.defaultValue = {};
      } else if (result.type === TYPE_BOOLEAN) {
        result.defaultValue = false;
      } else {
        result.defaultValue = null;
      }
    }
    return result;
  }

  _getConfigAttribute(props) {
    let attribute = 'value';
    if (props.multiple) {
      attribute += '[]';
    }
    let scalar = '?str';
    if (props.type === TYPE_BOOLEAN) {
      scalar = '?bool';
    } else if (props.type === TYPE_NUMBER) {
      scalar = '?num';
    } else if (props.type === TYPE_ASSOC) {
      scalar = '?id';
    } else if (props.type === TYPE_JSON) {
      scalar = '?json';
    }
    return attribute + scalar;
  }

  async getValue(key, props) {
    const confProps = this._getConfigProps({
      ...(CONFIG_PROPS[key] || {}),
      ...(props || {})
    });
    const attribute = this._getConfigAttribute(confProps);
    const currentValue = this._values[key] || {};
    if (currentValue.hasOwnProperty(attribute)) {
      return currentValue[attribute];
    }
    return this._loadConfigsFunc({ [key]: attribute }).then(atts => {
      return this.setValue(key, atts.result, confProps);
    });
  }

  async getValues() {
    for (let key in this._values) {
      await this._values[key];
    }
    return this._values;
  }

  setLoadConfigFunction(loadConfigFunc) {
    this._loadConfigsFunc = loadConfigFunc;
  }
}

window.Citeck.ConfigService = new ConfigService();
export default window.Citeck.ConfigService;
