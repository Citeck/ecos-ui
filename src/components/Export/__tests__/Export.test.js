import { render, screen } from '@testing-library/react';
import React from 'react';

import ConfigService from '../../../services/config/ConfigService';
import Export from '../Export';

describe('Export component tests', () => {
  test('should render Export component', () => {
    const spy = jest.spyOn(ConfigService, 'getValue').mockImplementation(() => Promise.resolve(true));

    const { container } = render(<Export />);

    expect(spy).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });

  test('should render Export component with empty className props', () => {
    const { container } = render(<Export />);
    expect(container.getElementsByClassName('ecos-btn-export')).toHaveLength(1);
  });

  test('should render Export component with className props', () => {
    const { container } = render(<Export className={'ecos-journal__settings-bar-export'} />);
    expect(container.getElementsByClassName('ecos-journal__settings-bar-export')).toHaveLength(1);
  });

  test('test equal dropdown with alfresco source and snapshots data', async () => {
    const spy = jest.spyOn(ConfigService, 'getValue').mockImplementation(() => Promise.resolve(true));

    const { container } = render(<Export />);

    expect(spy).toHaveBeenCalled();

    const dropdownSource = new Export().dropdownSourceVariants(true, false);
    const list = container.getElementsByTagName('ul').item(0);

    for (let node of Array.from(list.children)) {
      expect(node.textContent).toEqual(dropdownSource[dropdownSource.length - 1]['title']);
    }
  });

  test('test equal dropdown withoutAlfresco source and snapshots data', () => {
    const { container } = render(<Export />);

    const dropdownSource = new Export().dropdownSourceVariants(false, false);
    const list = container.getElementsByTagName('ul').item(0);

    for (let i = 0; i < list.length; i++) {
      expect(list[i].textContent).toEqual(dropdownSource[i]['title']);
    }
  });
});
