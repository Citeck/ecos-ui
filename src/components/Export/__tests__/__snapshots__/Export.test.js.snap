// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export component tests should render Export component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Export
    className=""
    dashletConfig={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Dropdown
          CustomItem={null}
          cascade={false}
          className=""
          controlClassName="ecos-btn_grey ecos-btn_settings-down ecos-btn_x-step_10"
          controlIcon="icon-download"
          direction="down"
          full={false}
          getStateOpen={[Function]}
          hasEmpty={true}
          hideSelected={false}
          isButton={true}
          isLinks={false}
          isStatic={true}
          labelIsDiv={false}
          menuClassName=""
          onChange={[Function]}
          placeholder=""
          right={false}
          scrollbarHeightMax="100%"
          scrollbarHeightMin="100%"
          source={
            Array [
              Object {
                "download": false,
                "id": 0,
                "target": "_blank",
                "title": "export-component.action.html-read",
                "type": "html",
              },
              Object {
                "download": true,
                "id": 1,
                "target": "_self",
                "title": "export-component.action.html-load",
                "type": "html",
              },
              Object {
                "download": true,
                "id": 2,
                "target": "_self",
                "title": "Excel",
                "type": "xlsx",
              },
              Object {
                "download": true,
                "id": 3,
                "target": "_self",
                "title": "CSV",
                "type": "csv",
              },
              Object {
                "click": [Function],
                "id": 4,
                "title": "export-component.action.copy-link",
              },
            ]
          }
          titleField="title"
          toggleClassName=""
          valueField="id"
          withScrollbar={false}
        />,
        <form
          encType="multipart/form-data"
          method="post"
        >
          <input
            name="jsondata"
            type="hidden"
            value=""
          />
        </form>,
      ],
      "className": "ecos-btn-export",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "CustomItem": null,
          "cascade": false,
          "children": undefined,
          "className": "",
          "controlClassName": "ecos-btn_grey ecos-btn_settings-down ecos-btn_x-step_10",
          "controlIcon": "icon-download",
          "direction": "down",
          "full": false,
          "getStateOpen": [Function],
          "hasEmpty": true,
          "hideSelected": false,
          "isButton": true,
          "isLinks": false,
          "isStatic": true,
          "labelIsDiv": false,
          "menuClassName": "",
          "onChange": [Function],
          "placeholder": "",
          "right": false,
          "scrollbarHeightMax": "100%",
          "scrollbarHeightMin": "100%",
          "source": Array [
            Object {
              "download": false,
              "id": 0,
              "target": "_blank",
              "title": "export-component.action.html-read",
              "type": "html",
            },
            Object {
              "download": true,
              "id": 1,
              "target": "_self",
              "title": "export-component.action.html-load",
              "type": "html",
            },
            Object {
              "download": true,
              "id": 2,
              "target": "_self",
              "title": "Excel",
              "type": "xlsx",
            },
            Object {
              "download": true,
              "id": 3,
              "target": "_self",
              "title": "CSV",
              "type": "csv",
            },
            Object {
              "click": [Function],
              "id": 4,
              "title": "export-component.action.copy-link",
            },
          ],
          "titleField": "title",
          "toggleClassName": "",
          "valueField": "id",
          "withScrollbar": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <input
            name="jsondata"
            type="hidden"
            value=""
          />,
          "encType": "multipart/form-data",
          "method": "post",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "jsondata",
            "type": "hidden",
            "value": "",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": "input",
        },
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Dropdown
            CustomItem={null}
            cascade={false}
            className=""
            controlClassName="ecos-btn_grey ecos-btn_settings-down ecos-btn_x-step_10"
            controlIcon="icon-download"
            direction="down"
            full={false}
            getStateOpen={[Function]}
            hasEmpty={true}
            hideSelected={false}
            isButton={true}
            isLinks={false}
            isStatic={true}
            labelIsDiv={false}
            menuClassName=""
            onChange={[Function]}
            placeholder=""
            right={false}
            scrollbarHeightMax="100%"
            scrollbarHeightMin="100%"
            source={
              Array [
                Object {
                  "download": false,
                  "id": 0,
                  "target": "_blank",
                  "title": "export-component.action.html-read",
                  "type": "html",
                },
                Object {
                  "download": true,
                  "id": 1,
                  "target": "_self",
                  "title": "export-component.action.html-load",
                  "type": "html",
                },
                Object {
                  "download": true,
                  "id": 2,
                  "target": "_self",
                  "title": "Excel",
                  "type": "xlsx",
                },
                Object {
                  "download": true,
                  "id": 3,
                  "target": "_self",
                  "title": "CSV",
                  "type": "csv",
                },
                Object {
                  "click": [Function],
                  "id": 4,
                  "title": "export-component.action.copy-link",
                },
              ]
            }
            titleField="title"
            toggleClassName=""
            valueField="id"
            withScrollbar={false}
          />,
          <form
            encType="multipart/form-data"
            method="post"
          >
            <input
              name="jsondata"
              type="hidden"
              value=""
            />
          </form>,
        ],
        "className": "ecos-btn-export",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "CustomItem": null,
            "cascade": false,
            "children": undefined,
            "className": "",
            "controlClassName": "ecos-btn_grey ecos-btn_settings-down ecos-btn_x-step_10",
            "controlIcon": "icon-download",
            "direction": "down",
            "full": false,
            "getStateOpen": [Function],
            "hasEmpty": true,
            "hideSelected": false,
            "isButton": true,
            "isLinks": false,
            "isStatic": true,
            "labelIsDiv": false,
            "menuClassName": "",
            "onChange": [Function],
            "placeholder": "",
            "right": false,
            "scrollbarHeightMax": "100%",
            "scrollbarHeightMin": "100%",
            "source": Array [
              Object {
                "download": false,
                "id": 0,
                "target": "_blank",
                "title": "export-component.action.html-read",
                "type": "html",
              },
              Object {
                "download": true,
                "id": 1,
                "target": "_self",
                "title": "export-component.action.html-load",
                "type": "html",
              },
              Object {
                "download": true,
                "id": 2,
                "target": "_self",
                "title": "Excel",
                "type": "xlsx",
              },
              Object {
                "download": true,
                "id": 3,
                "target": "_self",
                "title": "CSV",
                "type": "csv",
              },
              Object {
                "click": [Function],
                "id": 4,
                "title": "export-component.action.copy-link",
              },
            ],
            "titleField": "title",
            "toggleClassName": "",
            "valueField": "id",
            "withScrollbar": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <input
              name="jsondata"
              type="hidden"
              value=""
            />,
            "encType": "multipart/form-data",
            "method": "post",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "jsondata",
              "type": "hidden",
              "value": "",
            },
            "ref": Object {
              "current": null,
            },
            "rendered": null,
            "type": "input",
          },
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
